### ISSUE TITLE
Targeting multiple configurations with a single run

### ISSUE THREAD

**Issue by traceon** [conan-io/conan#799]
This is the first thing that always comes to mind when CMake generators that support multiple configurations (Debug vs Release, etc.; for example, Xcode and Visual Studio) are discussed. This doesn't give a solution to that problem, but definitely a step towards it.
In general, if any option, config, profile argument is a list of values and not a single value, all possible (if available/make sense) combinations of them can be downloaded/built, either by supplying that diversity to the native package build system, or by just building all the differently configured packages.
Does this make sense?

**Comment by lasote** [comment]
There is a work in progress that allows multi configuration at the issue #799. In the latest comments @memsharded explain how to use it, if you could give feedback would be great!

**Comment by traceon** [comment]
@lasote , #799 is this one. Did you mean some other issue?

**Comment by lasote** [comment]
Sorry. I meant #330

**Comment by memsharded** [comment]
I thought it was a duplicate, but actually it is decoupled from the CMake multi-generators, and could also be a use case for other systems, as the setup a developer might want to run in a new machine for their projects, even in linux/gcc. So I propose addressing it here, and if #330 will get into the release, I'll try to make this too. Thanks for the suggestion!

**Comment by lasote** [comment]
But it has a problem with regular generators, if I specify `cmake` or `gcc` or `txt` or any other distinct to the new `cmake_multi`... what conan should do? what setting combination will use to generate the file/s from the specified generator?

**Comment by memsharded** [comment]
Yes, I was probably wrong. This feature seems that could only make sense:

- With ``cmake_multi`` generator
- With a predefined subfolder creation pattern.

Regarding the second option, it is very likely that writing something generic enough into conan might be difficult, while writing some script in the user side is totally trivial:

```bash
$ mkdir mybuild1 && cd mybuild1
$ conan install .. -s build_type=Debug ....
$ cd .. && mkdir mybuild2 && cd mybuild2
$ conan install .. -s build_type=Release ....
```

In the same spirit, even if using the ``cmake_multi`` generator, it totally makes sense to have the explicit one-by-one install:

```bash
$ conan install .. -s build_type=Debug ....
$ conan install .. -s build_type=Release ....
```

My main concern is **inconsistency**, something that you will probably won't notice, but that we, the maintainers will. If the ``-s build_type=Debug,Release`` syntax is allowed, then users not using the ``cmake_multi`` generator will have unexpected behaviors, and then ask for support. 

Also, what would be the intended explosion of ``-s build_type=Debug,Release -s compiler.runtime=MDd,MD`` ? Maybe ``Debug-MDd`` and ``Release-MD``. But what happens if the user writes ``-s build_type=Debug,Release -s compiler.version=10,12,14``?

I propose to invoke the zen of python "explicit is better than implicit", especially when there are so many open issues and inconsistencies, and the explicit expansion of the combinations to be installed from the user side is so trivial.

**Comment by traceon** [comment]
@memsharded,

> But what happens if the user writes `-s build_type=Debug,Release -s compiler.version=10,12,14`?

The initial and yet the only way I was thinking about the behavior, as suggested in the original post, is the cartesian product of all options. And there is always profiles support, that can be used for fine grained configs.

@lasote, I was suggesting this as a general approach, orthogonal to cmake_multi.

As for CMake, I even think that having imported targets (only!) makes thing much more consistent, and this is even considered more modern CMake'ing (I understand that you are bound with backward compatibility/conservative tool support, but this is what I really choose for my projects that don't have to support people with ancient tools.)

As for generators, the prefixed/suffixed output files may be generated for each combination (and separately, an output file with regular name, for backward compatibility/default/fast usage.)

Apart from generator pitfall, there is another one - communicating multiple choices to a underlying package builders - some of which may support, for example, building and including both, debug and release versions in a single installation. Although, this is more of a bonus feature, not a showstopper for the original proposal.

**Comment by memsharded** [comment]
Yes, that is the point, the cartesian product of ``-s build_type=Debug,Release -s compiler.runtime=MDd,MD`` will create configurations that most likely will be wrong. And that would be the most necessary and used configuration by large.

The only place this split could make sense while keeping consistent would be for the ``--profiles`` argument.

**Comment by memsharded** [comment]
>As for CMake, I even think that having imported targets (only!) makes thing much more consistent, and this is even considered more modern CMake'ing (I understand that you are bound with backward compatibility/conservative tool support, but this is what I really choose for my projects that don't have to support people with ancient tools.)

Sorry, I don't understand this, could you please clarify? Are you talking about the current approach to modern CMake with targets? Maybe this feedback should be better moved to https://github.com/conan-io/conan/issues/601, where more interested people can give feedback. Also, if you haven't done it yet, please read that thread.

> As for generators, the prefixed/suffixed output files may be generated for each combination (and separately, an output file with regular name, for backward compatibility/default/fast usage.)

Same for this, I don't fully understand it. Might be moved to issue https://github.com/conan-io/conan/issues/330, or PR (for next release!): https://github.com/conan-io/conan/pull/804 ? It sounds like the implemented approach in the PR, could you please check it?

Thanks again for your feedback!

**Comment by traceon** [comment]
> Are you talking about the current approach to modern CMake with targets?

Yes, exactly, it's "Transitive Usage Requirements", #601 is about that. Thanks for the link.

> It sounds like the implemented approach in the PR, could you please check it?

Yes, looks like that's it. Thank you!

**Comment by traceon** [comment]
> the cartesian product of -s build_type=Debug,Release -s compiler.runtime=MDd,MD will create configurations that most likely will be wrong

Right, but it is not something unexpected: you specify wrong combination - you get wrong results.

**Comment by memsharded** [comment]
> Right, but it is not something unexpected: you specify wrong combination - you get wrong results.

Yes, but that is the main issue with it. The feature, if defined as cartesian product is unusable for the main purpose it was defined, installing at the same time Release & Debug configurations.

If instead of the cartesian product, we define it as the "inner product":

```
-s build_type=Debug,Release -s compiler.runtime=MDd,MD
# would be equal to 
-s build_type=Debug -s compiler.runtime=MDd
-s build_type=Release -s compiler.runtime=MD
```
That would work for that use case, but then it will be unusable for:
```
-s build_type=Debug,Release -s compiler.version=10,12,14
```
Then we will get user complains and bug reports that this is not working and a request to fix it so it works. I am pretty sure that if we define the "," expansion, this will happen.

That is why I could only understand the feature for ``profiles``, but not for settings.

**Comment by traceon** [comment]
Isn't the semantics of `build_type` defined separately for each library? So, in general, the `compiler.runtime` should be deduced from it.
I suspect that this particular example causes the confusion.

**Comment by memsharded** [comment]
Nop, it is a setting as the others, so the user can set it to any value, even if wrong. It is not deduced from other settings. Conan tries to follow the rule of being the less possible intrusive, opinionated and "automagic". 

If you can set any runtime (MD, MDd, MT, MTd) in the Visual Studio GUI, or in CMake, then, conan will never be "smart" and try to automatically deduce it. 

This was the rationale for doing this, from experience we know that this usually leads to block some users that are doing unconventional things, and want to be free to do what they want.

A recent mistake we did with this was assuming that within a dependency tree, all packages will necessarily have to share the same settings (as having the same compiler). We were wrong, and it was necessary to allow per-package settings, because users where mixing packages built with MinGW and with Visual Studio. If you haven't dont it yet, you could read (http://blog.conan.io/2016/12/22/Building-multi-compiler-C_and_C++-applications-with-conan.html). This somewhat reinforce the idea of not controlling the VS runtime, but allowing it to be freely set by the users.

**Comment by traceon** [comment]
Right, exactly. What I actually meant, is that `build_type` is just some setting, that usually implicitly  set several other settings at the library's build script level, so specifying `compiler.runtime=MDd` and `build_type=Debug` is something that shouldn't be done and shouldn't be recommended by any lib build system.

I.e., in any typical Visual Studio project, when you choose Debug configuration, you don't need to go and check MDd vs. MD options, moreover, you are strongly recommended against it.

**Comment by memsharded** [comment]
I see, you are right. 

The thing is that is not easy to achieve the same functionality in the command line. Even if you are discouraged, you are allowed, so it seems that conan should allow to do it. I wonder if we could make something without breaking existing behavior. If no command line argument is provided, then automatically choose the correct one? Then the default value defined in conan.conf would be discarded? What if defined in profiles? We'll have a look.

I guess the recommended way anyway would be to user ``profiles``, managing the settings explicitly is tedious and error prone.

**Comment by traceon** [comment]
I think it is an advantage that you can set any option, and I don't see any point in Conan doing analytics on compiler options.

I personally get irritated when suggested to use something like `build-type=debug` and knowing that it will set something implicitly, behind the scenes, and you'll need to enable verbose logging to see what flags are passed to compiler etc.
In contrast, it is very convenient to have something like `profile=debug` or `preset=debug` or `default=debug`, which just "includes" the appropriate config profile -- you always know that it's not a magic but some well defined config of the same common format is used, and you may just override this by your custom settings, and eventually you can go and see all the options it sets, and even compare `debug` vs `optimized` side by side, for example.

> I guess the recommended way anyway would be to user profiles, managing the settings explicitly is tedious and error prone.

Exactly

**Comment by sixten-hilborn** [comment]
If you're looking for a convenient way to use Conan with CMake you could have a look at [conan.cmake](https://github.com/conan-io/cmake-conan).
It will invoke `conan install ...` with settings from CMake automagically. If the settings aren't extracted as you would like you can just modify conan.cmake the way you need it.

**Comment by memsharded** [comment]
I think this issue has been addressed by conan 0.20 and multi-config packages, docs here: http://docs.conan.io/en/latest/packaging/package_info.html

Also with other approaches, like the ``cmake_multi`` generator or the ``build_id()`` method. All of these will also be summarized in a blog post we are drafting for this week.

So closing this issue, but please give feedback or reopen if necessary. Thanks!
